### ISSUE TITLE
Use HTTP basic auth credentials by default when provided

### ISSUE THREAD

**Issue by Hwesta** [httplib2/httplib2#16]
When `add_credentials` has been called, use the credentials in the initial request instead of expecting a 401 response before retrying with credentials.

If the credentials have been specified, it's reasonable to assume they're required and wouldn't have been provided if they weren't.  The current behavior causes many unnecessary requests that are know to fail. Not all servers return 401 on failed auth. I ran into problems where the 401 caused a broken pipe error, and the correctly authenticated call was never made.  Other examples exist in the [google code version of this bug](http://web.archive.org/web/20160111202154/https://code.google.com/p/httplib2/issues/detail?id=130)

For people looking for a workaround (from [this blog post](https://blog.josephscott.org/2011/06/07/http-basic-auth-with-httplib2/)), explicitly put the auth in the header initially.
```
import base64
import httplib2

h = httplib2.Http()
auth = base64.encodestring( 'username' + ':' + 'password' )

resp, content = h.request(
    'http://www.google.com/',
    'GET',
    headers = { 'Authorization' : 'Basic ' + auth }
)
```

**Comment by temoto** [comment]
I agree that there should be an option to send credentials immediately. But it should have an on/off trigger, per session and per request.

**Comment by rauberdaniel** [comment]
👍 This definitely took me some time to figure out. I’m still not sure why it didn’t work as the API returns a 401 when authentication failed. However, appending the auth header manually (as mentioned above) did work.

**Comment by galvez** [comment]
@httplib2/maintainers I'll be taking on this.
